Extended Tour Documentation (#614)
open-component-model/ocm: examples/lib/tour/02-composing-a-component-version/README.md

<!-- DO NOT MODIFY                                             -->
<!-- this file is generated by mdref                           -->
<!-- from ../docsrc/02-composing-a-component-version/README.md -->
# Composing a Component Version

This tour illustrates the basic usage of the API to
create/compose component versions.

It covers two basic scenarios:

> - [`basic`](01-basic-componentversion-creation.go) Create a component version stored in the filesystem

Reviewer: ```suggestion
- [`basic`](01-basic-componentversion-creation.go) Create a component version stored in the file system
```

Author: done

- [`compose`](02-composition-version.go) Create a component version stored in memory using a non-persistent composition version.

## Running the example

You can just call the main program with the scenario as argument. Configuration is not required.

## Walkthrough

### Basic Component Version Creation

The first variant just creates a new component version
in an OCM repository. To avoid the requirement for 
credentials a filesystem based repository is created, using
the *Common Transport Format* (CTF).

As usual, we start with getting access to an OCM context

> object

Reviewer: ```suggestion
object:
```

Author: done

```go
	ctx := ocm.DefaultContext()
```

To compose and store a new component version

> we finally need some OCM repository to

Reviewer: ```suggestion
we need some OCM repository to
```

Author: done

> store the component, The most simple

Reviewer: ```suggestion
store the component. The most simple
```

> external repository could be the filesystem.

Reviewer: ```suggestion
external repository could be the file system.
```

Author: done

> OCM defines a distribution format, the

Reviewer: ```suggestion
For this purpose OCM defines a distribution format,
```

Author: done

> Common Transport Format (CTF) for this,

Reviewer: ```suggestion
the Common Transport Format (CTF),
```

Author: done

which is an extension of the OCI distribution
specification.

> There are three flavours, *Directory*, *Tar* or *TGZ*.

Reviewer: ```suggestion
There are three flavors, *Directory*, *Tar* or *TGZ*.
```

Author: done

The implementation provides a regular OCM repository

> interface, like used in the previous example.

Reviewer: ```suggestion
interface, like the one used in the previous example.
```

Author: done

```go
	repo, err := ctfocm.Open(ctx, ctfocm.ACC_WRITABLE|ctfocm.ACC_CREATE, "/tmp/example02.ctf", 0o0744, ctfocm.FormatDirectory)
	if err != nil {
		return errors.Wrapf(err, "cannot create transport repository")
	}
	defer repo.Close()
```

Once we have a repository we can compose a new version.
First we create a new version backed by this repository.

> The result is a memory based representation not yet persisted.

Reviewer: ```suggestion
The result is a memory based representation, not yet persisted.
```

Author: done

```go
	cv, err := repo.NewComponentVersion(name, version)
	if err != nil {
		return errors.Wrapf(err, "cannot create new version")
	}
	defer cv.Close()
```

> Now, we can configure the component version. It exists in memory, only,

Reviewer: ```suggestion
Now, we can configure the component version. It only exists in memory so far,
```

Author: done

> so far, but is already connected to the repository.

Reviewer: ```suggestion
but is already connected to the repository.
```

Author: done

The setup of the component version is put into a 
separate method (`setupVersion`), so it can be reused for the second variant.

First, we configure the component version provider.

```go
	provider := &compdesc.Provider{
		Name: "acme.org",
	}
	fmt.Printf("  setting provider...\n")
	err := cv.SetProvider(provider)
	if err != nil {
		return errors.Wrapf(err, "cannot set provider")
	}
```

> The provider is a structure with a name and some labels, we just set

Reviewer: ```suggestion
The provider is a structure with a name and some labels,
```

> the name, here.

Reviewer: we just set `Name` here.

> the name, here.

Reviewer: ```suggestion
we just set `Name` here.
```

Author: done

Now, we fill the component version with content.

> First, we add some resource already located in

Reviewer: ```suggestion
First, we add a resource that is already located in
```

> some external registry. We use an OCI image here.

Reviewer: ```suggestion
an external registry. We use an OCI image here.
```

Author: done

A resources has some metadata, like an identity
and a type.
The identity is just a set of string properties,

> with at least containing the name property.

Reviewer: ```suggestion
at least containing the `name` property.
```

Author: done

Additional identity properties can be added via
options.
The type represents the logical meaning of the
resource, here an `ociImage`.

```go
	meta, err := elements.ResourceMeta("image", resourcetypes.OCI_IMAGE)
	if err != nil {
		// without metadata options, there will be never be an error,
		// bit to be complete, we just handle the error case, here.
		return errors.Wrapf(err, "invalid resource meta")
	}
```

> In this example, we just use the name property.

Reviewer: ```suggestion
In this example, we just use the `name` property.
```

Author: done

> And most important a resource requires content.

Reviewer: ```suggestion
And most importantly, a resource requires content.
```

Author: done

> Content can be already present in some external

Reviewer: ```suggestion
Content can already be present in some external
```

Author: done

repository. As long, as there is an access type
for this kind of repository, we can just refer to it.
Here, we just use an image provided by the
OCM ecosystem.
Supported access types can be found under
.../pkg/contexts/ocm/accessmethods.

```go
	acc := ociartifact.New("ghcr.io/open-component-model/ocm/ocm.software/toi/installers/helminstaller/helminstaller:0.4.0")
```

Once we have both, the metadata and the content specification,

> we can now add the resource. 

Reviewer: ```suggestion
we can add the resource. 
```

Author: done

The `SetResource` methods will replace an existing resource with the same
identity, or add the resource, if no such resource exists in the component
version.

```go
	err = cv.SetResource(meta, acc)
	if err != nil {
		return errors.Wrapf(err, "cannot add access to ocmcli-image)")
	}
```

Now, we will add a second resource, some unspecific yaml data.
Therefore, we use the generic YAML resource type.
In practice, you should always use a resource type describing
the real meaning of the content, for example something like

> `kubernetesManifest`, This enables tools working with specific content

Reviewer: ```suggestion
`kubernetesManifest`. This enables tools working with specific content
```

Author: done

to understand the resource set of a component version.

```go
	meta, err = elements.ResourceMeta("descriptor", resourcetypes.OCM_YAML)
	if err != nil {
		return errors.Wrapf(err, "invalid resource meta")
	}
```

Besides referring to external resources, another possibility
to add content is to directly provide the content blob. The
used abstraction here is `blobaccess.BlobAccess`.
Any blob content provided by an implementation of this
interface can be added as resource.
There are various access implementations for blobs

> taken from the local host, for example, from the filesystem,

Reviewer: kannst du mal schauen, ob du das etwas einfacher formuliert bekommst? for example außerhalb der Klammer, dann ein weiteres for example innerhalb der Klammer. Vielleicht zwei Sätze oder das zweite Beispiel weglassen?

Author: done

or from other repositories (for example by mapping
an access type specification into a blob access).
The most simple form is to directly provide a byte sequence,
for example some YAML data.
A blob always must provide a mime type, describing the
technical format of the blob's byte sequence. This is different

> from the resource type. A logical resource, like a helm chart can be

Reviewer: ```suggestion
from the resource type. A logical resource, like a Helm chart can be
```

Author: done

> represented in different technical formats, for example a helm chart

Reviewer: ```suggestion
represented in different technical formats, for example a Helm chart
```

Author: done

archive or as OCI image archive. While the type described the
logical content, the meaning of the resource, its mime type
described the technical blob format used to represent
the resource as byte sequence.

```go
		blob := blobaccess.ForString(mime.MIME_YAML, yamldata)
```

When storing the blob, it is possible to provide some
optional additional information:
- a name of the resource described by the blob, which could
  be used to do a later upload into an external repository
  (for example the image repository of an OCI image stored
  as local blob)
- an additional access type, which provides an alternative

>   global technology specific access to the same content.

Reviewer: ```suggestion
  global technology specific access
```

Author: done

>   we don't use it, here.

Reviewer: ```suggestion
 to the same content (we don't that in this example).
```

Author: done

```go
		err = cv.SetResourceBlob(meta, blob, "", nil)
		if err != nil {
			return errors.Wrapf(err, "cannot add yaml document")
		}
```

Resources added by blobs will be stored along with the component
version metadata in the same repository, no external
repository is required.

The above blob example describes the basic operations,
which can be used to compose any kind of resource 
from any kind of source.

> For selected use cases there are convenience helpers,

Reviewer: ```suggestion
For selected use cases, there are convenience helpers available,
```

Author: done

which can be used to compose a resource access object.
This is basically the same interface returned by `GetResource`
functions on the component version from the last example.
Such objects can directly be used to add/modify a resource in a
component version.

> The above case could be written as follows, also:

Reviewer: ```suggestion

The above case could also be written as follows:
```

```go
		res := textblob.ResourceAccess(cv.GetContext(), meta, yamldata,
			textblob.WithimeType(mime.MIME_YAML))
		err = cv.SetResourceAccess(res)
		if err != nil {
			return errors.Wrapf(err, "cannot add yaml document")
		}
```

The resource access is an abstraction of external access via access
methods or direct blob access objects and additionally
contain all the required resource metadata.

There are even more complex blob sources, for example

> for helm charts stored in the filesystem, or even for images

Author: done

generated by docker builds.
Here, we just compose a multi-platform image built with buildx
from these sources (components/ocmcli) featuring two flavors.
(you have to execute `make image.multi` in components/ocmcli
before executing this example.

```go
	meta, err = elements.ResourceMeta("ocmcli", resourcetypes.OCI_IMAGE)
	if err != nil {
		return errors.Wrapf(err, "invalid resource meta")
	}
	res := dockermultiblob.ResourceAccess(cv.GetContext(), meta,
		dockermultiblob.WithPrinter(common.StdoutPrinter),
		dockermultiblob.WithHint("ocm.software/ocmci"),
		dockermultiblob.WithVersion(current_version),
		dockermultiblob.WithVariants(
			fmt.Sprintf("ocmcli-image:%s-linux-amd64", current_version),
			fmt.Sprintf("ocmcli-image:%s-linux-arm64", current_version),
		),
	)
	err = cv.SetResourceAccess(res)
	if err != nil {
		return errors.Wrapf(err, "cannot add ocmcli")
	}
```

### Composition Environment

The second variant just creates a new component version
in a memory based composition environment, no persistence is 
required. Like all component versions, such component versions

> can be added to any repository, later.

Reviewer: ```suggestion
can be added to any repository later.
```

As usual, we start with getting access to an OCM context
object

```go
	ctx := ocm.DefaultContext()
```

Now, we can create a new component version in the composition

> environment. It does not require a repository or component object.

Author: done

```go
	cv := composition.NewComponentVersion(ctx, "acme.org/example2", "v0.1.0")
```

To configure the component version, we can just reuse the coding
from the example above, the component version interface is just the same.

> We just call the `setupversion` function on the created component version access.

Reviewer: ```suggestion
We just call the `setupVersion` function on the created component version access.
```

Author: done

```go
	err := setupVersion(cv)
	if err != nil {
		return errors.Wrapf(err, "version composition")
	}
```

The resulting component version can be added to any OCM repository,

> like gthe onw from the previous example.

Reviewer: ```suggestion
like the one from the previous example.
```

Author: done
